src/main/java/com/epam/esm/service/GiftCertificateService.java

```diff
@@ -0,0 +1,133 @@
+    }
+
+    public GiftCertificateDto getGiftCertificate(long id) throws EntityNotExistsException {
+        Optional<GiftCertificate> optionalGiftCertificate = giftCertificateDao.findById(id);
+        if (optionalGiftCertificate.isEmpty()) {
+            throw new EntityNotExistsException();
+        }
+
+        return dtoMapper.map(optionalGiftCertificate.get());
+    }
+
+    @Transactional
+    public void deleteEntity(long id) throws EntityNotExistsException {
+        if (giftCertificateDao.findById(id).isEmpty()) {
+            throw new EntityNotExistsException();
+        }
+
+        giftCertificateTagDao.deleteGiftCertificateById(id);
+        giftCertificateDao.deleteById(id);
+    }
+
+    @Transactional
+    public GiftCertificateDto updateGiftCertificate(long id, GiftCertificateDto giftCertificateDto) throws EntityNotExistsException, BadEntityException {
+        if (giftCertificateDao.findById(id).isEmpty()) {
+            throw new EntityNotExistsException();
+        }
+
+        GiftCertificate giftCertificate = dtoMapper.unmap(giftCertificateDto);
+
+        if (!validator.validateUpdating(giftCertificate)) {
+            throw new BadEntityException();
+        }
+
+        giftCertificate.setLastUpdateDate(DATE_FORMAT.format(new Date()));
+
+        checkForTags(giftCertificate.getTags());
+        giftCertificateDao.updateGiftCertificate(id, giftCertificate);
+
+        if (!giftCertificate.getTags().isEmpty()) {
+            connectCertificatesAndTags(id, giftCertificate);
+        }
+        return getGiftCertificate(id);
+    }
+
+    @Transactional
+    public GiftCertificateDto createGiftCertificate(GiftCertificateDto giftCertificateDto) throws BadEntityException, EntityNotExistsException {
+        GiftCertificate giftCertificate = dtoMapper.unmap(giftCertificateDto);
+
+        if (!validator.validateCreating(giftCertificate)) {
+            throw new BadEntityException();
+        }
+
+        giftCertificate.setCreationDate(DATE_FORMAT.format(new Date()));
+        giftCertificate.setLastUpdateDate(DATE_FORMAT.format(new Date()));
+
+        checkForTags(giftCertificate.getTags());
+        long lastInsertedId = giftCertificateDao.create(giftCertificate);
+        giftCertificateTagDao.createConnections(lastInsertedId, giftCertificate.getTags());
+
+        return getGiftCertificate(lastInsertedId);
+    }
+
+    private void checkForTags(Set<Tag> newTags) {
+        List<String> newTagNames = new ArrayList<>();
+        newTags.stream().forEach(tag -> newTagNames.add(tag.getName()));
+        newTagNames.stream().filter(tagName -> tagDao.findTagByName(tagName).isEmpty()).forEach(tagName -> tagDao.create(new Tag(tagName)));
+    }
+
+    private void connectCertificatesAndTags(long id, GiftCertificate giftCertificate) {
+        List<Long> tagIdsBeforeUpdate = giftCertificateTagDao.getIdsBeforeUpdate(id);
+        List<Long> tagIdsAfterUpdate = tagDao.getIdsAfterUpdate(giftCertificate.getTags());
```

Comment:
I assume that by this time `giftCertificate.getTags()` should return you a fully initalized set of tags with tag id's. So why do we need to call this method?

Reply:
After we implemented initializeTags() properly method we do not need the giftCertificate.getTags() at all